### ISSUE TITLE
How to use platinum-searcher (pt) as the default search tool?

### ISSUE THREAD

**Issue by tavurth** [hlissner/doom-emacs#2800]
**What are you trying to achieve?**
---
I'd like to get my doom mode using [the_platinum_searcher](https://github.com/monochromegane/the_platinum_searcher) as it's default search tool (projectile project search), as I've found it more robust than the doom default.

**What have you tried?**
---
I'm not sure where to start, as this seems like something which could be a cool feature in Doom

**Additional information**
---
Try a projectile text search `M-m s f`, and on large projects pt seems to not lose references. Have been having trouble in some projects finding all instances of a keyword.

**System information**
---
<details>

```
emacs   version    26.3
        features   NOTIFY ACL GNUTLS LIBXML2 ZLIB TOOLKIT_SCROLL_BARS NS MODULES THREADS
        build      Sep 02, 2019
        buildopts  (--with-ns '--enable-locallisppath=/Library/Application Support/Emacs/${version}/site-lisp:/Library/Application Support/Emacs/site-lisp' --with-modules)
        windowsys  ns
        daemonp    server-running
doom    version    2.0.9
        build      HEAD -> develop, origin/develop, origin/HEAD 21db1b7c9 2020-02-13 14:45:08 -0500
        dir        ~/.doom.d/
system  type       darwin
        config     x86_64-apple-darwin18.2.0
        shell      /bin/zsh
        uname      Darwin 19.3.0 Darwin Kernel Version 19.3.0: Thu Jan  9 20:58:23 PST 2020; root:xnu-6153.81.5~1/RELEASE_X86_64 x86_64
        path       (~/.nvm/versions/node/v12.16.0/bin /usr/local/bin /usr/bin /bin /usr/sbin /sbin /Library/TeX/texbin /Applications/Emacs.app/Contents/MacOS/bin-x86_64-10_14 /Applications/Emacs.app/Contents/MacOS/libexec-x86_64-10_14 /Applications/Emacs.app/Contents/MacOS/libexec/)
config  envfile    envvar-file
        elc-files  0
        modules    (:completion company ivy :ui doom doom-dashboard hl-todo modeline ophints (popup +all +defaults) vc-gutter workspaces :editor file-templates (format +onsave) multiple-cursors rotate-text snippets :emacs dired electric ibuffer vc :checkers syntax :tools (eval +overlay) (lookup +docsets) magit :lang data emacs-lisp javascript markdown (org +dragndrop +present) python sh :config (default +bindings +smartparens))
        packages   (n/a)
        elpa       (glsl-mode)
        unpin      (n/a)
```
</details>

**Comment by hlissner** [comment]
Doom does not support other search tools. It is built around ripgrep (and is in fact one of Doom's hard dependencies) because its featureset, compared to ag and pt are effectively the same and to reduce maintenance costs. I have no intention of expanding support. I'd be curious to know how pt is more robust.

Some possibilities that may help:

1. You could try changing `counsel-rg-base-command` to match `counsel-pt-base-command`.
2. Since you have the `:completion ivy` module enabled, `counsel-pt` is available to you. You can invoke it directly: `M-x counsel-pt`, or bind it to keys of your choosing:

   ```elisp
   (map! :leader "sp" #'counsel-pt)
   ```
3. If the UI is the problem, then perhaps [deadgrep](https://github.com/Wilfred/deadgrep) is a better answer than switching to pt.

**Comment by tavurth** [comment]
> Doom does not support other search tools. It is built around ripgrep (and is in fact one of Doom's hard dependencies) because its featureset (and performance), compared to ag and pt, are effectively identical (if not superior) and to reduce maintenance costs. I have no intention of expanding support. 

Interesting about `ripgrep` dependency, was sure I'd seen it mentioned before but couldn't find the post.

> I'd be curious to know how pt is more robust.
I'm a fairly new doom user, around 4-5 months, but I've been using emacs in some form for the last 10 years. I've been using pt around 2-3 of those years.

pt's caching system seems to be very fast to update. For `ripgrep` I've noticed sometimes my queries don't give me an up to date view of the search structure, example below:

When I was searching today for something in my codebase, I made a `M-m s p` (search in project), for a specific search string, and then a `C-c C-e` (edit this search buffer). 

After editing and saving, my project was broken, after some investigation I found that `ripgrep` had not correctly shown me all the variants for editing, and so I was missing around 3 critical changes.

> Some possibilities that may help: 
You could try changing counsel-rg-base-command to match counsel-pt-base-command.
Since you have the :completion ivy module enabled, counsel-pt is available to you. You can invoke it directly: M-x counsel-pt, or bind it to keys of your choosing:

Thank you for the suggestions! I will look into them tomorrow

**Comment by hlissner** [comment]
> pt's caching system seems to be very fast to update. For ripgrep I've noticed sometimes my queries don't give me an up to date view of the search structure

Can't say I've ever experienced this, despite heavy use. What version of ripgrep are you using?

**Comment by tavurth** [comment]
> What version of ripgrep are you using?

I'm not sure, how can I get that information? (Don't see it in doom info)
I'm using whichever ripgrep the latest doom version uses.

**Comment by tavurth** [comment]
Unfortunately I'm still seeing this issue often :(

**Comment by BurntSushi** [comment]
> pt's caching system seems to be very fast to update. For `ripgrep` I've noticed sometimes my queries don't give me an up to date view of the search structure, example below:

Neither pt nor ripgrep have any explicit caching system.

If you can reproduce the missing matches via a normal `rg` command on a corpus we can both access, I'd be happy to receive a bug report. My guess is that ripgrep is ignoring files by respecting your `.gitignore`. To check this, try your search with `-uuu`. If ripgrep still doesn't find what you expect, then perhaps there is a bug. If ripgrep does find it, then re-run the search without `-uuu` but with `--debug`. That should tell you why some files aren't being searched.

**Comment by tavurth** [comment]
@BurntSushi @hlissner Thanks for getting back to me.

I posted yesterday thinking there was still a problem with the `ripgrep` implementation, however I removed my post when I discovered it was a `projectile` issue. 

Although it would be nice to support different base search utilities, I will close this issue for now as I cannot reproduce the issue I was having before. 

If I can reproduce it I will open this again. I will also attempt to make a repo for testing in that case 👍
